File: pyntk/ntk/lib/log.py
##
# This file is part of Netsukuku
# (c) Copyright 2009 Daniele Tricoli aka Eriol <eriol@mornie.org>
#
# This source code is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as published
# by the Free Software Foundation; either version 2 of the License,
# or (at your option) any later version.
#
# This source code is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# Please refer to the GNU Public License for more details.
#
# You should have received a copy of the GNU Public License along with
# this source code; if not, write to:
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
##

import logging
import logging.handlers
import os.path
import sys
import traceback

from ntk.config import settings

LOG_FILE = os.path.join(settings.LOG_DIR, settings.LOG_FILE)

logger = logging.getLogger('')

def init_logger():
    ''' Configure the logging system using `settings'. '''

    global logger
    if settings.VERBOSE_LEVEL > 0:
        m = ('%(asctime)s %(levelname)s:'
             '(%(filename)s at line %(lineno)d): %(message)s')
    else:
        m = '%(asctime)s %(levelname)s: %(message)s'

    formatter = logging.Formatter(m)

    rfh = logging.handlers.RotatingFileHandler(LOG_FILE,
                                               maxBytes=1024 * 20,
                                               backupCount=8)
    rfh.setFormatter(formatter)

    console = logging.StreamHandler()
    console.setFormatter(formatter)

    logger.ULTRADEBUG = 5
    logging.addLevelName(logger.ULTRADEBUG, 'ULTRADEBUG')
    if settings.VERBOSE_LEVEL > 4: settings.VERBOSE_LEVEL = 4
    levels = {0 : logging.ERROR, 1 : logging.WARNING, 2 : logging.INFO, 
              3 : logging.DEBUG, 4 : logger.ULTRADEBUG}
    logger.setLevel(levels[settings.VERBOSE_LEVEL])
    logger.addHandler(rfh)

    if settings.DEBUG_ON_SCREEN:
        logger.addHandler(console)

def get_stackframes(back=0):
    ret = sys._current_frames().items()[0][1]
    while ret is not None and ret.f_back and back >= 0:
        ret = ret.f_back
        back -= 1
    return get_stackframes_repr(ret)

def get_stackframes_repr(frame):
    ret = []
    while True:
        ret.append((frame.f_code.co_filename, frame.f_code.co_name, 
                    frame.f_lineno))
        frame = frame.f_back
        if not frame: break
    return ret.__repr__()

def log_exception_stacktrace(e, indent=2):
    spaces = ' ' * indent
    excinfo = sys.exc_info()
    tb = excinfo[2]
    logger.error(spaces + "Exception: %s" % (e.__repr__(), ))
    logger.error(spaces + "Stacktrace:")
    frames = traceback.extract_tb(tb)
    for fr in frames:
        logger.error(spaces + "  File \"%s\", line %s, in %s" % (fr[0], 
                                                                 fr[1], 
                                                                 fr[2]))
        logger.error(spaces + "    %s" % (fr[3], ))

def log_on_file(filename, msg):
    f = None
    try:
        f = open(filename, 'a')
    except:
        return
    try:
        f.write(str(msg) + '\n')
    finally:
        f.close()

logger.log_on_file = log_on_file

File: src/log.c
/* This file is part of Netsukuku
 * (c) Copyright 2004 Andrea Lo Pumo aka AlpT <alpt@freaknet.org>
 *
 * This source code is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Public License as published 
 * by the Free Software Foundation; either version 2 of the License,
 * or (at your option) any later version.
 *
 * This source code is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * Please refer to the GNU Public License for more details.
 *
 * You should have received a copy of the GNU Public License along with
 * this source code; if not, write to:
 * Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <stdio.h>
#include <syslog.h>
#include <stdarg.h>

#include "log.h"

char *__argv0;
int dbg_lvl;
int log_to_stderr;
static log_facility=LOG_DAEMON;

void log_init(char *prog, int dbg, int log_stderr)
{
	__argv0=prog;
	dbg_lvl=dbg;
	log_to_stderr=log_stderr;
}

/* Life is fatal! */
void fatal(const char *fmt,...)
{
	char str[strlen(fmt)+3];
	va_list args;

	str[0]='!';
	str[1]=' ';
	strncpy(str+2, fmt, strlen(fmt));
	str[strlen(fmt)+2]=0;

	va_start(args, fmt);
	print_log(LOG_CRIT, str, args);
	va_end(args);
	exit(1);
	/*TODO: safe_exit(255); We must save the maps before exit*/
}

/* Misc errors */
void error(const char *fmt,...)
{
	char str[strlen(fmt)+3];
	va_list args;

	str[0]='*';
	str[1]=' ';
	strncpy(str+2, fmt, strlen(fmt));
	str[strlen(fmt)+2]=0;
	
	va_start(args, fmt);
	print_log(LOG_ERR, str, args);
	va_end(args);
}

/* Let's give some news */
void loginfo(const char *fmt,...)
{
	char str[strlen(fmt)+3];
	va_list args;

	str[0]='+';
	str[1]=' ';
	strncpy(str+2, fmt, strlen(fmt));
	str[strlen(fmt)+2]=0;
	
	va_start(args, fmt);
	print_log(LOG_INFO, str, args);
	va_end(args);
}

/* "Debugging is twice as hard as writing the code in the first place.
 * Therefore, if you write the code as cleverly as possible, you are,
 * by definition, not smart enough to debug it." - Brian W. Kernighan
 * Damn!
 */

void debug(int lvl, const char *fmt,...)
{
	char str[strlen(fmt)+3];
	va_list args;

	if(lvl <= dbg_lvl) {
		str[0]='#';
		str[1]=' ';
		strncpy(str+2, fmt, strlen(fmt));
		str[strlen(fmt)+2]=0;

		va_start(args, fmt);
		print_log(LOG_DEBUG, str, args);
		va_end(args);
	}
}

void print_log(int level, const char *fmt, va_list args)
{
	
	if(log_to_stderr) {
		vfprintf(stderr, fmt, args);
		fprintf(stderr, "\n");
	} else {
		openlog(__argv0, LOG_PID, log_facility);
		vsyslog(level | log_facility, fmt, args);
		closelog();
	}
}
